packages/compose/src/hooks/use-merge-refs/index.js

```diff
@@ -0,0 +1,42 @@
+/**
+ * WordPress dependencies
+ */
+import { useRef, useCallback } from '@wordpress/element';
+
+/** @typedef {import('@wordpress/element').RefObject} RefObject */
+/** @typedef {import('@wordpress/element').RefCallback} RefCallback */
+
+/**
+ * Merges refs into one ref callback. Ensures the merged ref callbacks are only
+ * called when it changes or when the ref value changes. If you don't wish a ref
+ * callback to be called on every render, wrap it with `useCallback( ref, [] )`.
+ * Dependencies can be added, but when a dependency changes, the ref callback
+ * will be called with the same node.
+ *
+ * @param {Array<RefObject|RefCallback>} refs The refs to be merged.
+ */
+export default function useMergeRefs( refs ) {
+	const lastValue = useRef( null );
+	const lastRefs = useRef( refs );
+
+	return useCallback( ( value ) => {
+		refs.forEach( ( ref, index ) => {
+			if ( typeof ref === 'function' ) {
+				// Only call a ref when it wants to be called:
+				// - The value changes.
+				// - The ref callback has changed (e.g. an updated dependency).
+				if (
+					value !== lastValue.current ||
```

Comment:
I believe comparing to the previous value is not useful because if one of the ref  changes, React will first call the old function with "null" and then call the new  function with the new  node which  mean this check is always true which also means that the implementation here is equivalent to `useCallback( merge(refs),  refs )` and in both situations  we can't have a callback with dependencies as ref because if it changes, all the other merged refs, even the ones that didn't change,w ill be called again (creating memory leaks or unexpected behavior)

Reply:
> all the other merged refs, even the ones that didn't change,w ill be called again

No, they won't be called since we check if the callback has been changed. Only if it's changed we call it, and we leave any others that remain the same alone. The value check is there to ensure that, even if the callback didn't change, the ref callback is still called.

Reply:
Added clearer documentation in ece1982. I tested ref callbacks with and without dependencies and they both work as expected. If a dependency changes, old callback is called with `null`, the new with the node. If no dependency changes, the callback is never called, even if other ref callback dependencies change.

Reply:
I'll try to add some test cases tomorrow.

Reply:
Here's the failing test case:

```
const ref1 = useCallback( dosomething, [dep1]);
const ref2 = useCallback( dosomething2, [] );
const ref = useMergeRef( [ ref1, ref2 ] );

if the node doesn't change and dep1 change, `dosomething2` will be called again while it shouldn't be. (I didn't test but that's my read of the code) Let's add this unit test and see.

Reply:
You're right, it's calling the wrong callback. Changed the code and started adding tests, but it's still a work in progress.